src/main/java/decodes/cwms/CwmsConnectionPool.java

```diff
@@ -0,0 +1,399 @@
+    {
+        CwmsConnectionPool ret = pools.get(info);
+        if (ret == null)
+        {
+            try(Connection conn = pool.getConnection(info.getLoginInfo());)
+			{
+                ret = new CwmsConnectionPool(info,conn);
+                pools.put(info,ret);        
+            }
+            catch(SQLException ex)
+            {
+                throw new BadConnectException("Unable to initialize pool.",ex);
+            }
+            catch(DbIoException ex)
+            {
+                throw new BadConnectException("Unable to set office context.",ex);
+            }
+        }
+        return ret;
+    }
+
+
+    private CwmsConnectionPool(CwmsConnectionInfo info, Connection conn) throws BadConnectException,SQLException,DbIoException
+    {
+        this.info = info;
+        String officeId = info.getLoginInfo().getUserOfficeId();
+        String username = info.getLoginInfo().getUser();
+        info.setConnection(conn);
+        info.setDbOfficeCode(officeId2code(conn, officeId));
+
+        // MJM 2018-2/21 Force autoCommit on.
+        try{ conn.setAutoCommit(true); }
+        catch(SQLException ex)
+        {
+            log.warning("Cannot set SQL AutoCommit to true: " + ex);
+        }
+
+        String priv = getOfficePrivileges(conn, info.getLoginInfo().getUrl(), officeId);
+        setCtxDbOfficeId(conn, officeId, info.getDbOfficeCode(), priv);
+
+
+        try
+		{
+            String name = String.format("CwmsConnectionPool(%s/%s)",info.getLoginInfo().getUrl(),info.getLoginInfo().getUser());
+			ManagementFactory.getPlatformMBeanServer()
+							 .registerMBean(this, new ObjectName("org.opendcs:type=TimeSeriesDb,name=\""+name+"\""));
+		}
+		catch(JMException ex)
+		{
+			log.warning("Unable to register tracking bean " + ex.getLocalizedMessage());
+		}
+
+    }    
+
+    @Override
+	public int getConnectionsOut()
+    {
+		return connectionsOut.size();
+	}
+
+	@Override
+	public int getConnectionsAvailable()
+    {
+		return 10 - connectionsOut.size();
+	}
+
+	@Override
+	public String getThreadName()
+    {
+		return Thread.currentThread().getName();
+	}
+
+	@Override
+	public int getGetConnCalled()
+    {
+		return this.connectionsRequested;
+	}
+
+	@Override
+	public int getFreeConnCalled()
+    {
+		return this.connectionsFreed;
+	}
+
+	@Override
+	public int getUnknownReturned()
+    {
+		return this.unknownConnReturned;
+	}
+
+	@Override
+	public TabularData getConnectionsList() throws OpenDataException
+    {
+		TabularData td = new TabularDataSupport(JMXTypes.CONNECTION_LIST);
+        for(WrappedConnection conn: this.connectionsOut)
+        {
+            td.put(conn.asCompositeData());
+        }
+        return td;
+	}
+
+
+    public Connection getConnection() throws SQLException
+    {
+        connectionsRequested++;
+        Connection conn = pool.getConnection(info.getLoginInfo());
+        WrappedConnection wc = new WrappedConnection(conn,(c)->CwmsDbConnectionPool.close(c),Optional.of(this));
```

Comment:
Shouldn't this call `returnConnection(c)` instead of passing to the `CwmsDbConnectionPool`?

Reply:
No, it's operating on the actual connection.... granted that would've worked to and would clean up some logic I just implemented.

Reply:
thank you, that simplified things a lot.